Refactor the script

dak180 opened the pull request:
Switch to bash (was already using bashisms)
Update syntax.
Fix shellcheck warnings.

- [x]  Fixes #5.
- [x]  Fixes #4.

[Tarball](https://github.com/fink/passwd/files/12074721/passwd-20230717.tgz) for testing (see fink/fink-distributions#1031 for updated info files).

dak180 committed:
First pass.

Switch to bash (was already using bashisms)
Update syntax.
Fix shellcheck warnings.

dak180 committed:
A few more fixups.

dak180 committed:
Rename files.

dak180 committed:
More fixups.

dak180 committed:
Rewrite the script from the ground up.

dak180 committed:
Use AutoUid by default.

dak180 committed:
Standardize replacement values.

dak180 committed:
Ignore build products.

dak180 committed:
Add the make file.

dak180 committed:
Fixups.

dak180 committed:
Add the mailmap file.

dak180 committed:
More fixups.

dak180 committed:
Fix the help message.

dak180 committed:
Add the man page.

dak180 committed:
Fix plurals.

dak180 committed:
Fix AUTHORS.txt requirements.

dak180 committed:
Include built man page in the tarball.

Reduces the need to have a2x installed for installation.

dak180 commented:
@fink/fink-developers the primary things to to test are that reboots are not required to use the newly created user with `chown` and that the users do not show up in the fast user switching menu. You should be able to `make install` from the repo and then just use it from the command line to test; the Directory Utility app can be used to inspect the resulting users.

dak180 committed:
Update the news file.

dak180 committed:
Be sure to setup the group before the user.

dak180 committed:
Set (or lookup) the GID properly.

dak180 committed:
Be better about user visible feedback.

dak180 committed:
More news updates.

Be more explicit about users (or groups) being reused if the only difference in name is a leading '_'.

dak180 committed:
Add a note.

dak180 committed:
Add some comments.

dak180 commented:
[passwd-20160421.tgz](https://github.com/fink/passwd/files/12016597/passwd-20160421.tgz) is a curent tarball to test with.

dak180 committed:
Add a sha256 checksum.

TheSin- commented on update-passwd.tool:
```diff
@@ -0,0 +1,370 @@
+
+
+# Setup group
+echo "Checking to see if the grooup ${GROUPNAME} exists:"
```
group not grooup

dak180 committed:
Hide error output.

Also fix spelling.

dak180 commented on update-passwd.tool:
```diff
@@ -0,0 +1,370 @@
+
+# Setup group
+echo "Checking to see if the grooup ${GROUPNAME} exists:"
+if dscl . -read "/groups/${GROUPNAME}" PrimaryGroupID; then
```
The `if` there is not dependent on the command output but rather the status; any status other than `0` should return false (a missing entry should return a status of 56).

TheSin- replied:
I agree and it outputs 0 on error ;)

dak180 replied:
Interesting; if I run `dscl . -read "/groups/_redis" PrimaryGroupID; echo $?` here I get:
```
<dscl_cmd> DS Error: -14136 (eDSRecordNotFound)
56
```

and you get something like:
```
<dscl_cmd> DS Error: -14136 (eDSRecordNotFound)
0
```
? 

I hate not being able to count on exit status.

TheSin- replied:
no I get what I posted in IRC

```
dscl . -read "/groups/_redis" PrimaryGroupID
No such key: PrimaryGroupID
echo $?
0
```

dak180 replied:
And especially interesting since I get the same output on `uname -r` = `22.5.0`, `21.6.0`, `20.6.0`, and `16.7.0`.

TheSin- replied:
yeah and this system is the same as you.

The test system is fully vanilla, nothing at all installed except XCode and XDarwin, though nothing should affect those commands that I know of. I'll do a little investigation tomorrow though it might need to wait until Thursday, but it's very odd

TheSin- replied:
kinda a far shot but could it be shell? I know on this system I switched it to bash, the other test system would be default which makes me nuts.  Trying to think of nothing, but program output shouldn't be dependant on shell :\

dak180 replied:
`No such key: PrimaryGroupID` makes it seem like `dscl` is not working right but why/how?

TheSin- replied:
the test system is on an MDM, maybe that changes something?  But it's still something we need to plan for.  Hopefully I can figure it out.  Cause I can't see that being an issue either.

dak180 replied:
Actually that may be the issue if it means that it is not using the local user directory.  I would be very interested in what nodes are listed in the Directory Utility app (/System/Library/CoreServices/Applications).

dak180 replied:
@TheSin- what do you get when running `dscacheutil -configuration` on the mdm system?

TheSin- replied:
```
% sudo dscacheutil -configuration
DirectoryService Cache search policy:
    /Local/Default

Unable to get details from the cache node
Unable to get cache configuration information
```

```
% sudo dscacheutil -q group -a name clamav; echo $?
name: _clamav
password: *
gid: 82

0
```

```
% sudo dscacheutil -q group -a name _redis; echo $?
0
```

dak180 replied:
I am going to switch over to `dscacheutil` since that seems to give consistent output across all systems. Just to double check though `id` does not give you any weirdness, right?

TheSin- replied:
dscacheutil still gives 0 when it's missing though, unless you are checking the output and not the exit code now.

dak180 replied:
> so far id seems okay, anything specific you want me to test?

I just did it gives sane exit codes for both missing and present users.

TheSin- commented:
yeah and this system is the same as you.

The test system is fully vanilla, nothing at all installed except XCode and XDarwin, though nothing should affect those commands that I know of.  I'll do a little investigation tomorrow though it might need to wait until Thursday, but it's very odd

dak180 committed:
Use `dscacheutil` instead of `dscl` to test for existence.

`dscl` may have inconsistent results when combined with MDM.

dak180 committed:
Skip when existing group found.

TheSin- commented:
```
Setting up passwd-redis (20230529-1) ...
Checking to see if the group redis exists:
I could not find an unused GID in the range 600 - 699.
You can expand this range by changing the AutoUidMin and/or
AutoUidMax values in /opt/sw/etc/passwd.conf.

##########################################

Make sure to run 'fink reinstall passwd-redis'
to install the redis entry, or you can create
it manually if you would prefer to do that.

##########################################

Group not found.
Checking to see if the user redis exists:
redis does not exist; creating...
2023-07-13 09:34:25.555 sysadminctl[12807:56642] User named 'redis' already exists.
id: redis: no such user
```

TheSin- commented:
```
~ % id 600                     
uid=600(fink-bld) gid=20(staff) groups=20(staff),600(fink-bld),12(everyone),61(localaccounts),701(com.apple.sharepoint.group.1),98(_lpadmin),100(_lpoperator)
~ % id 601
id: 601: no such user
```

TheSin- commented:
must not be walking right?

```
~ % dscacheutil -q group -a gid 600; echo $?
name: fink-bld
password: *
gid: 600
users: fink-bld 

0
~ % dscacheutil -q group -a gid 601; echo $?
0
```

dak180 committed:
Be more explicit about operation for dseditgroup.

Also fix a typo.

dak180 committed:
Exit if group or user creation fails.

TheSin- commented:
With the latest changes

```
Setting up passwd-redis (20230529-1) ...
Checking to see if the group redis exists:
Create called on existing record - do you want to overwrite, y or n : y

Checking to see if the user redis exists:
redis does not exist; creating...
2023-07-15 11:04:27.610 sysadminctl[56726:330808] User named 'redis' already exists.
id: redis: no such user
```

```
Setting up passwd-redis (20230529-1) ...
Checking to see if the group redis exists:
redis exists.
Checking to see if the user redis exists:
redis does not exist; creating...
2023-07-15 11:05:05.226 sysadminctl[56863:331750] User named 'redis' already exists.
id: redis: no such user
```

so group looks good now!  But user still failing.

dak180 commented:
@TheSin- glad to hear that group is working.  For the user stuff did you check in the Directory Utility app to see if there is any partially created users that it running into (`id` is not reliable for doing this)?

TheSin- commented:
<img width="809" alt="Screenshot 2023-07-16 at 12 07 04 PM" src="https://github.com/fink/passwd/assets/1868661/52ed2396-b10a-4973-9f8b-964eddab1c3f">
<img width="810" alt="Screenshot 2023-07-16 at 12 07 16 PM" src="https://github.com/fink/passwd/assets/1868661/2be12c1e-48c5-414f-b786-a8f53e850c00">

TheSin- commented:
so looks like there is a user even then id says no ?

TheSin- commented:
deleted the user in Directory Utility and then re ran the install

```
Setting up passwd-redis (20230529-1) ...
Checking to see if the group redis exists:
redis exists.
Checking to see if the user redis exists:
redis does not exist; creating...
2023-07-16 12:09:03.225 sysadminctl[59828:578275] ----------------------------
2023-07-16 12:09:03.226 sysadminctl[59828:578275] No clear text password or interactive option was specified (adduser, change/reset password will not allow user to use FDE) !
2023-07-16 12:09:03.226 sysadminctl[59828:578275] ----------------------------
2023-07-16 12:09:03.233 sysadminctl[59828:578275] New account password error.  (5402)
id: redis: no such user
```

dak180 committed:
Try single quotes.

dak180 commented:
@TheSin- try changing the id range in the config file to between 200 and 400.

dak180 committed:
Role Accounts require an id in the range 200-400 and a name starting with _.

This is more strictly enforced in later versions of macos.

dak180 committed:
Use dsimport for user creation.

dsimport has less usage restrictions but is easier to shoot yourself in the foot with.

dak180 committed:
Add quotes.

dak180 committed:
Clean up user visible output.

dak180 committed:
Conform to package versioning.

TheSin- commented:
With the current changes

```
Setting up passwd-redis (20230529-1) ...
Checking to see if the group redis exists:
redis exists.
Checking to see if the user redis exists:
redis does not exist; creating...
redis
name: redis
password: *
uid: 601
gid: 601
dir: /opt/sw/var/db/redis
shell: /usr/bin/false
gecos: Redis Key-Value Store Server
```

```
% id -u redis
601
```

```
% id -g redis
601
```

```
Setting up passwd-postgres (20230529-1) ...
Checking to see if the group postgres exists:
postgres exists.
Checking to see if the user postgres exists:
postgres exists.
```

dak180 committed:
Better user messages.

dak180 committed:
Flush the cache.

TheSin- commented:
I'm not sure the group is getting set right? I'd assume the primary group for Redis would be Redis

```
% mkdir blah                    
% sudo chown redis: blah
% ls -lad blah                  
drwxr-xr-x  2 redis  staff  64 18 Jul 07:33 blah
```

But I was able to use it without a reboot, I'm going to try and build and install Redis now to make sure.

TheSin- commented:
According to Directory Utility Redis's primaryGroupID is 601 so odd `chown redis:` doesn't do `redis:redis`.  But it seems right just odd behaviour from chown I think.

redis built and installed without a reboot!!!

I do not see new users in the fast switching menu, I'm going to reboot now to make sure.

Was there anything else to test?

dak180 commented:
> I'm not sure the group is getting set right?

I would assume that since you (as a normal user) ran `mkdir` that is where `staff` got set and since you did not specify a change of group, it stayed the same.

TheSin- reviewed (approved):
tested and looking perfect to me.

This is very important for 10.13+ and should be adopted immediately IMHO.  We should also use this code during bootstrapping for `fink-bld`.

TheSin- commented:
> > I'm not sure the group is getting set right?
> 
> I would assume that since you (as a normal user) ran `mkdir` that is where `staff` got set and since you did not specify a change of group, it stayed the same.

then that is my own ignorance with `chown`, I thought :<blank> defaulted to the user group for chown not the current console user.  Makes sense though.  So that's on me and things are working 100%
